File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/renderer/level/BlockSelectorRenderer.java
package pize.tests.voxelgame.client.renderer.level;

import pize.app.Disposable;
import pize.files.Resource;
import pize.graphics.util.Shader;
import pize.math.vecmath.matrix.Matrix4f;
import pize.math.vecmath.vector.Vec3f;
import pize.tests.voxelgame.client.block.shape.BlockCursor;
import pize.tests.voxelgame.client.control.camera.GameCamera;
import pize.tests.voxelgame.client.control.BlockRayCast;

public class BlockSelectorRenderer implements Disposable{
    
    private final LevelRenderer levelRenderer;
    private final Shader shader;
    private final Matrix4f translationMatrix;
    
    public BlockSelectorRenderer(LevelRenderer levelRenderer){
        this.levelRenderer = levelRenderer;
        
        this.shader = new Shader(new Resource("shader/line.vert"), new Resource("shader/line.frag"));
        
        this.translationMatrix = new Matrix4f();
    }
    
    
    public void render(GameCamera camera){
        final BlockRayCast rayCast = levelRenderer.getGameRenderer().getSession().getGame().getBlockRayCast();
        if(!rayCast.isSelected())
            return;
        
        shader.bind();
        shader.setUniform("u_projection", camera.getProjection());
        shader.setUniform("u_view", camera.getView());
        
        translationMatrix.toTranslated(new Vec3f(rayCast.getSelectedBlockPosition()).sub(camera.getX(), 0, camera.getZ()));
        shader.setUniform("u_model", translationMatrix);
        
        final BlockCursor shape = rayCast.getSelectedBlockProps().getState(rayCast.getSelectedBlockState()).getCursor();
        if(shape != null)
            shape.render();
    }
    
    @Override
    public void dispose(){
        shader.dispose();
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/renderer/level/BlockSelectorRenderer.kt
package pize.tests.voxelgame.client.renderer.level

import pize.app.Disposable
import pize.files.Resource
import pize.graphics.util.Shader
import pize.math.vecmath.matrix.Matrix4f
import pize.math.vecmath.vector.Vec3f
import pize.tests.voxelgame.client.control.camera.GameCamera

class BlockSelectorRenderer(private val levelRenderer: LevelRenderer) : Disposable {
    private val shader: Shader
    private val translationMatrix: Matrix4f

    init {
        shader = Shader(Resource("shader/line.vert"), Resource("shader/line.frag"))
        translationMatrix = Matrix4f()
    }

    fun render(camera: GameCamera) {
        val rayCast = levelRenderer.gameRenderer.session.game.blockRayCast
        if (!rayCast!!.isSelected) return
        shader.bind()
        shader.setUniform("u_projection", camera.projection)
        shader.setUniform("u_view", camera.view)
        translationMatrix.toTranslated(Vec3f(rayCast.selectedBlockPosition).sub(camera.x, 0f, camera.z))
        shader.setUniform("u_model", translationMatrix)
        val shape = rayCast.selectedBlockProps.getState(rayCast.selectedBlockState.toInt()).cursor
        shape?.render()
    }

    override fun dispose() {
        shader.dispose()
    }
}
